File: BottomNavigationBar/src/main/java/studio/mandysa/bottomnavigationbar/BottomNavigationBar.java
package studio.mandysa.bottomnavigationbar;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.LinearLayout;

import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("ALL")
public class BottomNavigationBar extends LinearLayout implements OnClickListener {

    private final List<NavigationItem> mItemList = new ArrayList<>();
    private int mPosition;
    private OnItemViewSelectedListener mListener;

    public BottomNavigationBar(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public BottomNavigationBar setTextColor(int checked) {
        return setTextColor(checked, 0);
    }

    public BottomNavigationBar setTextColor(int checked, int unchecked) {
        for (NavigationItem item : mItemList) {
            item.setTextColor(checked, unchecked);
        }
        return this;
    }

    public void setOnItemViewSelectedListener(OnItemViewSelectedListener mListener) {
        this.mListener = mListener;
    }

    @Override
    public void onClick(View p1) {
        NavigationItem item = (NavigationItem) p1;
        if (mPosition != item.getPosition()) {
            ((NavigationItem) getChildAt(mPosition)).check();
            item.check();
            mPosition = item.getPosition();
        } else return;
        if (mListener != null)
            mListener.onclick(item.getPosition());
    }

    public void setPosition(int position) {
        onClick(getChildAt(position));
        if (mListener != null) mListener.onclick(position);
    }

    public int getPosition() {
        return mPosition;
    }

    public BottomNavigationBar addItem(String text) {
        addItem(0, text);
        return this;
    }

    public BottomNavigationBar addItem(int image) {
        addItem(image, null);
        return this;
    }

    public BottomNavigationBar addItem(int image, int textRes) {
        addItem(image, getResources().getString(textRes));
        return this;
    }

    public void addItem(int image, String text) {
        NavigationItem item = new NavigationItem(mItemList.size(), text, getContext(), image);
        item.setOnClickListener(this);
        mItemList.add(item);
        addView(item);
        if (mItemList.size() == 1) mItemList.get(0).check();
    }

    public interface OnItemViewSelectedListener {
        void onclick(int position);
    }

}

File: bottomnavigationbar/src/main/java/studio/mandysa/bottomnavigationbar/BottomNavigationBar.kt
package studio.mandysa.bottomnavigationbar

import android.content.Context
import android.graphics.PorterDuff
import android.util.AttributeSet
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.recyclerview.widget.RecyclerView
import mandysax.lifecycle.livedata.LiveData
import mandysax.lifecycle.livedata.MutableLiveData
import studio.mandysa.jiuwo.utils.recyclerAdapter
import studio.mandysa.jiuwo.utils.setup
import studio.mandysa.jiuwo.utils.staggered


class BottomNavigationBar : RecyclerView {

    @ColorInt
    private var mActiveColor = context.getColor(R.color.default_checked_color)

    @ColorInt
    private var mInActiveColor = context.getColor(R.color.default_unchecked_color)

    private val mSelectedPosition = MutableLiveData<Int>()

    fun setActiveColorResource(@ColorRes colorRes: Int): BottomNavigationBar {
        return setActiveColor(context.getColor(colorRes))
    }

    fun setActiveColor(@ColorInt color: Int): BottomNavigationBar {
        mActiveColor = color
        return this
    }

    fun setInActiveColorResource(@ColorRes colorRes: Int): BottomNavigationBar {
        return setInActiveColor(context.getColor(colorRes))
    }

    fun setInActiveColor(@ColorInt color: Int): BottomNavigationBar {
        mInActiveColor = color
        return this
    }

    fun getSelectedPosition(): LiveData<Int> {
        return mSelectedPosition
    }

    fun setSelectedPosition(position: Int) {
        mSelectedPosition.value = position
    }

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    var models: List<BottomNavigationItem>? = null
        set(value) {
            field = value
            staggered(field!!.size).setup {
                addType<BottomNavigationItem>(R.layout.item_vertical)
                onBind {
                    val model = getModel<BottomNavigationItem>()
                    val itemIcon = itemView.findViewById<ImageView>(R.id.image1)
                    val itemTitle = itemView.findViewById<TextView>(R.id.text1)
                    itemIcon.setImageResource(model.mImageRes)
                    itemIcon.setColorFilter(mInActiveColor)
                    itemTitle?.text = model.mText
                    itemTitle?.setTextColor(mInActiveColor)
                    itemView.setOnClickListener {
                        setSelectedPosition(modelPosition)
                    }
                    mSelectedPosition.observeForever { it ->
                        when (it == modelPosition) {
                            true -> {
                                if (model.mActiveColor == null) mActiveColor else model.mActiveColor!!
                            }
                            false -> {
                                if (model.mInActiveColor == null) mInActiveColor else model.mInActiveColor!!
                            }
                        }.apply {
                            itemIcon.setColorFilter(this, PorterDuff.Mode.SRC_IN)
                            itemTitle?.setTextColor(
                                this
                            )
                        }
                    }
                }
            }
            recyclerAdapter.models = field
        }

}